---
title: Getting Started with Thesis Writing in R
bibliography:
  - "../Templates/thesis-citations.bib"
  - ../Templates/packages.bib
csl: ../Templates/apa.csl
output:
  pdf_document:
    toc: false
    latex_engine: xelatex
    fig_caption: true
    includes:
      in_header: "../Templates/MtA-Thesis-Preamble.tex"
fontsize: 12pt
geometry:
- margin = 1in
linestretch: 1.5
classoption: oneside
indent: true
editor_options:
  chunk_output_type: inline
always_allow_html: true

# Use nocite to include any citations that aren't included in-text (i.e. R packages)
---

```{r set-global-options, echo=FALSE}
# ------------ Set chunk options --------------------------------------------------------------
# cairo_pdf encodes high-quality plots with searchable text
# Hide chunk warnings to reduce distractions
# comment = NA removes the double hash tags next to chunk output

knitr::opts_chunk$set(
  dev = c("cairo_pdf"),
  dpi = 600,
  fig.align = "center",
  comment = NA,
  echo = FALSE,
  message = FALSE,
  warning = FALSE
)
```

```{r load-libraries, include = FALSE}
# ------------- Load packages ----------------------------------------------------------------------------
# Here are some potentially useful packages to start off with - free feel to delete them if you don't need them
library(here)
library(dplyr) # Data manipulation
library(ggplot2) # Make plots
library(viridis) # Accessible and beautiful colour palettes
library(extrafont) # Required for custom fonts in plots
loadfonts()
library(ggtext) # Required for formatting plot text (e.g. coloured text in title)
library(patchwork) # Required for multi-plot layouts
library(ggsignif) # Required for significance stars and brackets on plots
library(ggh4x) # Required to create nested x-axis labels
library(ggpubr) # Required for significance brackets

# Generate a bibliography of all packages used in this file
# I will use this later to cite all packages that I used in my bibliography

#knitr::write_bib(file = here("Templates","packages.bib"))
```

```{r set-plot-options}
# ---------- Set plot theme, colours & sizes --------------------------------------------------------------------------------
# Set save_choice to "yes" if you want to save plots (! this will increase the run time; generally only run this after adding new data)
# To use this, set up an if/else statement with ggsave nested within it

save_choice <- "no"

# Set colours here only for consistency. Here is an example
# Refer to these colours in your ggplot code, for example geom_point(color = my_colours[2]) for blue points.

my_colours <- c("#6600cc", "#0093fb", "#55b323","#ffe70f","#e86c00","#333333", "#411900","#e11584")

# Custom fonts may cause issues depending on what fonts you have in your system
# Troubleshooting steps: 
  # Try changing the font to one that you have on your computer
  # If it does not work, you could always delete 'family = plot_font_family' in the ggplot theme set below
plot_font_family <- "Segoe UI"

# A consistent y-axis enables comparison across multiple experiments, treatments, etc.
# To use this, define variables like y_axis_limit and then use these as limits in coord_cartesian() in your ggplots.

y_axis_limit <- 175
```

```{r set-ggplot2-theme}
# ----------- Set ggplot theme ----------------------------------------------------------------------------------
# You should experiment to find a ggplot theme that works for your project
# Here is an example of defining a ggplot theme.

# This theme increases the axis font sizes and removes the default gray background.
# Requires the extrafont() package (loaded in the load-libraries chunk) for custom font

theme_set(
  theme_classic() %+replace%
    theme(
      text = element_text(
        family = plot_font_family
      ),
      plot.title = element_text(
        color = "black",
        size = 20,
        family = plot_font_family,
        margin = margin(b = 25),
        hjust = 0.5
      ),
      plot.margin = margin(25, 25, 25, 25),
      plot.caption = element_text(
        hjust = 0,
        family = plot_font_family,
        size = 12
      ),
      plot.caption.position = "plot",
      axis.text = element_text(
        size = 12,
        color = "black"
      ),
      axis.title = element_text(
        size = 16,
        face = "bold"
      ),
      axis.title.y = element_text(
        margin = margin(r = 25),
        angle = 90,
        vjust = 0.5
      ),
      axis.title.x = element_text(margin = margin(b = 25, t = 20)),
      axis.ticks = element_blank(),
      strip.background = element_rect(color = NA, fill = NA),
      strip.text = element_text(size = 20)
    )
)
```


\begin{centering}
\begin{spacing}{1}
\Large

by

\LARGE

Christelinda Laureijs

\end{spacing}
\end{centering}

\large
\vspace*{\fill}

<!-- Although it is not ideal, I am manually setting the linebreaks here. This is to conform to MtA's title page standards -->


\begin{centering}
\begin{spacing}{1}
A short guide for curious and motivated researchers

June 17, 2024

\end{spacing}
\end{centering}


\newpage
\vspace*{\fill}
\large


<!-- The preamble includes \usepackage{adforn}. Adforn enables you to add fancy decorative embellishments. Look up adforn on CTAN to see a list of available symbols. -->

\adforn{20} *COLOPHON* \adforn{48}

\normalsize

This document was typeset in R Markdown and the text was set in EB Garamond. The data and code used to produce this document are available at [*https://github.com/christelinda-laureijs/reproducible-thesis*](https://github.com/christelinda-laureijs/reproducible-thesis).

\newpage
\pagenumbering{roman}


\setcounter{tocdepth}{2}
\tableofcontents
\listoffigures

\cleardoublepage
\normalsize
\pagenumbering{arabic}


# Introduction

Writing a paper in R can be challenging at first, but it is a rewarding process that will save you lots of time, effort and stress over the long term. This document provides examples and sample code for you to get started. It is also a sample RMarkdown document to see if you can knit PDFs from R and RStudio. Ideally, you should be able to click on Knit -> Knit to PDF and generate a PDF without issues. 

*A note for Mac Users:* When you knit the document the first time, you might see error messages with something about Cairo and your graphics device. Every device is different, but you might be able to fix it by installing XQuartz and then Cairo. 

## What are the advantages of writing in R?

### Reproducibility

Your analyses and text are all in one place, so you will always know what you did to create a specific plot or how you got a certain p-value. Others will be able to follow your process by reading your code, and if you make your code and data availble, they can replicate your analyses (and ideally end up with the same numbers!).

### Efficiency

Rather than typing the same things over and over again, you can set up functions to complete tasks like completing a statistical test or creating a plot. Once your code is set up, it is easy to run your paper again with new data. This is especially helpful when you're frequently adding new data.

Since you can define your colours, fonts, and ggplot theme at the beginning of your document, you don't need to fiddle with formatting each plot. All plots will have the same formatting, and it is easy to change colours/styling for all plots at once.

### No manual formatting

You don't need to worry about the layout of your document at all. You do not need to manually number your figures or subheadings, format your Table of Contents, or readjust paragraphs each time you insert or remove a picture. You can easily cross-reference figures, equations and chapters. To insert a list of figures, just type \\listoffigures - see, it's very doable!

Since RMarkdown is a plain-text document, your computer will be able to handle large documents much more easily than Word. It will not get buggy or slow when you have lots of high-resolution pictures.

### Beautiful typography

When you knit a document using a PDF, R uses \LaTeX. \LaTeX is a typesetting engine that uses mathematical algorithms to arrange text into the most ideal way according to typesetting rules. The resulting documents look very elegant in a way that is hard to quantify. \LaTeX-produced documents look very distinctive and after a while, you'll be able to easily tell the differences between a document produced using \LaTeX vs. Word.

Even if you stick to just the defaults, your documents will look great. If you need to use equations, \LaTeX is one of the best ways to create clean, well-aligned equations.

### Free

R, Rmarkdown, \LaTeX and the packages you'll use are all free and open-source.

### Longevity

Your documents are plaintext files, which  means that they don't take up much space on your computer, and you can open them up years later in any plain text editor (even Notepad!). It also means that your files will always be available, and you won't get locked into proprietary software.

### Version Control

It is easy to set up version control with tools like Git and GitHub (which is also free). I would highly recommend learning how to set up a project on GitHub. Each time you make a change to the document on your computer, Git will also save the changes to the online version of your file, hosted on GitHub. It means that you'll always have a backup copy, and if something wrong happens, you can revert back to an older version and see what changes you made.


## What are the disadvantages of writing in R?

### Learning curve

If you're not familiar with R, RMarkdown, and \LaTeX, it will take much longer to set up your paper than in Word. It will also take a while to format your preamble to make your document look exactly the way you want it. You'll likely spend a lot of time googling things and reading through answers on StackOverflow and GitHub.

### Troubleshooting

Things will break down, and sometimes you'll spend way more time than you had anticipated dealing with error messages. During these times, it does seem much easier to just open a Word document and type there.

To minimize errors, I would suggest these things:

  - Run each code chunk from top to bottom to catch any errors. Ensure that everything runs before knitting.
  - Knit your document frequently as you go to help spot errors. 
  - R will knit documents using a blank R session, so everything that needs to be in the document must be defined in your script.
  - Turn off the "Save workspace image" option in your R Global Options to prevent old loaded packages and hidden variables from 'hovering' in the background and creating strange rrors.
  - Frequently use Run -> Restart R and clear output to prevent objects from cluttering your workspace and causing dependencies. For example, if you define a variable in the console, but not your document you won't realize the problem until you have a fresh R session. 
  - Use knitr::knit_exit() to stop knitting early. It can be a useful way to identify the specific line of code that is causing knitting issues.

### Separate content and layout

You don't get to see what your document looks like until you knit it. This can be disconcerting for some people but great for others because you aren't distracted by formatting. Although you should knit your document often, don't fuss around with formatting until your paper is almost done. Things like paragraph spacing will change, and it is better to wait until the end.

At the beginning, you may worry about floats (things like pictures, plots, and tables). \LaTeX will 'float' these over the text and then plop them down in a way that minimizes the number of paragraph breaks and blank spots. This means that your floats will often be further from where you want them. It is very, very difficult to 'force' floats to go into a specific spot, and the place that \LaTeX chooses is often the best layout-wise. Always use references like Figure 1, rather than "the figure below".

## Collaboration issues

You can't use track changes, comments, or shared files. The best way to simulate this is to set up a GitHub repository and add your supervisor as a collaborator. This may cause issues if you have a supervisor who doesn't know how to use GitHub and they don't want to comment a PDF.


# How-Tos

## Inserting citations

Here is an example of an in-text citation: Insulin is a hormone that regulates blood glucose levels, digestive processes, and body weight [@demeyts2000]. When you write [\@demeyts2000], the citation will be rendered as an in-text citation. This document uses APA formatting because of the 'apa.csl' file. If you want a different format, you can download the appropriate CSL file online.

## Inserting Plots

You can plot figures in R. This is one of the best parts - R will generate the plots each time the document knits so you don't have to repeatedly copy and paste pictures into your paper!

```{r}
#| fig.cap = "Here is an example of a figure caption",
#| fig.scap = "Sample plot in R",
#| out.width = "100%"


cars %>% 
  ggplot(aes(x = speed, y = dist)) +
  geom_point() +
  labs(x = "Speed", y = "Distance")
```

## Inserting Images

You can insert images through knitr. fig.cap can be quite long, and you can also add a label in the form of '\\\\label{text}' When you write a reference to Figure '\\\\label{insulin-pathway}', \LaTeX will print out Figure \ref{insulin-pathway}.

*Important:* To use `fig.cap` and `fig.scap`, you must define an `out.width` in the chunk options. If not, R will not recognize these as \LaTeX commands.

```{r insulin-schematic}
#| fig.cap = "When insulin binds to a cell, it activates a series of molecular pathways involved in energy metabolism and gene expression...[if your figure caption is too long, you should define a short figure caption (fig.scap) which will go into the list of figures]...although it looks like there are two distinct pathways, the downstream components of the P13K and MAPK pathways frequently interact with one another (De Meyts, 2000). \\label{insulin-pathway}",
#| fig.scap = "Insulin binding activates a series of molecular pathways.",
#| out.width = "100%",
#| purl = FALSE

# fig.scap = short caption that LaTeX will use in the List of Figures
# out.width = 100% of line width. Must define out.width OR use out.extra = '' to allow LaTeX to recognize this as LaTeX code.
# Excluded from Scripts-from-Thesis-Rmd.R (purl = FALSE) because this is for the thesis only

knitr::include_graphics(here("Figures/Insulin-schematic.png"))
```


## Changing the formatting

If you want to change how any part of this document looks, go to Templates/MtA-Thesis-Preamble.tex.

# References

```{r}
# Use knit_exit() anytime throughout your document to create a stop point. This is useful for if you only have the first part of your document finished or if you are troubleshooting.
knitr::knit_exit()
```

# R Review

Here are some chunks if you need a refresher with R.

```{r}
# You can use R as a calculator
2+3

3*5

2/5
```

```{r assign-variables}
# You can assign values to variables
# Make x equal to 2
x <- 2

# Use variables in equations
3*x

x + 2
```


```{r making-lists}
# Defining lists in R

number_list <- c(1, 9, 8, 3, 4, 2, 2)


# Use functions like mean()
mean(number_list)
max(number_list)
min(number_list)
```

## Importing data

The following sample code is for a project where the goal was to analyze changes in a neuron's current amplitude over time.

```{r}
# Import cell data from "Sample-data.csv"
# This comes from the folder Data/
# Header = True because I already have named my columns

raw_df <- read.csv(here("Data/Sample-data.csv"), header = T)
```

```{r}
# Get a preview of the first few rows of the data
head(raw_df)
```

```{r}
# Clampex records negative values for currents
# Negative transform P1 and P2

raw_df <- raw_df %>%
  mutate(
    P1 = P1*-1,
    P2 = P2*-1
  )
```

## Filter data

```{r filtering}
# Show cells from male rats only

raw_df %>%
  filter(Sex == "Male")
```

## Dealing with errors

R is very picky about typos, upper-case vs. lower-case, and other details. What are the mistakes below?

```{r errors, eval=F}
# What is the mistake below?
raw_df %>%
   filter(sex == "Male")
```

```{r, eval=F}
# What is the problem below?
raw_df %>%
  filter(Sex = "Female")
```

```{r, eval=F}
# What is wrong with the code below?
raw_df %>%
  filter(Sex == "Female"')

```

## Plotting

The goal is to make a scatterplot of P1 vs. Time for various cells. First, I am showing an example of plotting filtering just one recording of interest. This has the ID "L".

```{r}
# Filter one recording of interest
cell_L <- raw_df %>%
  filter(Letter == "L")
```

```{r}
# Plot cell L only
# P1 is on the y-axis
# Time is on the x-axis
# Use geom_point() because this is a scatterplot

ggplot(cell_L, aes(x = Time, y = P1)) +
  geom_point()
```
To add customizations to your plot, you can keep adding new layers. The first layer that I'm adding here is the labs() layer to make more informative axis titles.

```{r}
# Example: adding better axis labels
ggplot(cell_L, aes(x = Time, y = P1)) +
  geom_point() +
  labs(
    x = "Time (min)",
    y = "eEPSC amplitude (pA)"
  )
```

I recorded cells in 5-minute intervals. I included `color = ID` to colour each recording separately. This is just to show an example of how you can include additional variables in a plot.

```{r}
# Color by ID
ggplot(cell_L, aes(x = Time, y = P1, color = ID)) +
  geom_point() +
  theme_classic()
```

You can use `ggplot` to make beautiful plots very easily. Here are some examples that show some of ggplot's capabilities.

The plot below shows the data for all of my cells on the same plot. I used `color = Letter` to make each cell a different colour.


```{r}
# Plot all cells
# Use a different color for each recording

ggplot(raw_df, aes(x = Time, y = P1, color = Letter)) +
  geom_point() +
  theme_minimal() +
  theme(legend.position = "none") # hide legend
```

In Excel and Prism, it is very difficult to make faceted plots. R makes this very simple. Here is a faceted plot showing all of the cells from male rats with the PPP treatment applied. It is *much* easier to use facet plots instead of creating individual scatterplots for each letter.

```{r}
raw_df %>%
  filter(Sex == "Female" & Treatment == "PPP") %>%
  ggplot(aes(x = Time, y = P1, color = Letter)) +
  geom_point() +
  theme_minimal() +
  theme(legend.position = "none") +
  facet_wrap(.~Letter, ncol = 3)
```

You can use R to make art!

```{r}
# Artistic example!
raw_df %>%
  filter(Time <=30) %>%
  ggplot(aes(x = Time, y = P1, color = Letter)) +
  geom_point(alpha = 0.6, size = 1.4) +
  theme_void() +
  theme(legend.position = "none")
```
## Practice

If you want some extra practice, try changing the code and seeing what happens. You can also try these questions.

1a. Create a variable called "y" and assign it to a value of 25.
1b. Use a function to calculate the square root of y.

2a. Filter the data to only include cells where I applied a certain treatment.
2b. Filter the data to only include cells from male rats in the control condition. (Hint: Use the & symbol to connect two conditions).

3a. Make a plot of P1 vs. Time for recording "AA".
3b. Make a plot of P1 vs. Time for female rats only.
3c. Make a plot of P1 vs. Time for PPP-treated cells only, and color it by cell.
